### PR TITLE
builder: return image ID in API when using buildkit

### PR DESCRIPTION
**Pull request by tiborvass**
Signed-off-by: Tibor Vass <tibor@docker.com>

### CODE REVIEW

**Comment by tiborvass** [comment]
Ping @tonistiigi @AkihiroSuda

File: builder/builder-next/builder.go [hunk 1]
```diff
@@ -225,6 +225,29 @@ func (b *Builder) Build(ctx context.Context, opt backend.BuildConfig) (*builder.
 
 	eg, ctx := errgroup.WithContext(ctx)
 
+	// TODO: shouldn't this be using pkg/streamformatter instead?
```
Comment (tonistiigi) [inline builder/builder-next/builder.go]: yes, can we just add a type parameter to `streamformatter.AuxFormatter`

File: builder/builder-next/builder.go [hunk 2]
```diff
@@ -235,7 +258,7 @@ func (b *Builder) Build(ctx context.Context, opt backend.BuildConfig) (*builder.
 			return errors.Errorf("missing image id")
 		}
 		out.ImageID = id
-		return nil
+		return writeAux("moby.image.id", struct{ ID string }{id})
```
Comment (tonistiigi) [inline builder/builder-next/builder.go]: This probably doesn't need a struct

**Comment by tiborvass** [comment]
@tonistiigi @AkihiroSuda updated

File: api/server/backend/build/backend.go [hunk 3]
```diff
@@ -72,8 +72,11 @@ func (b *Backend) Build(ctx context.Context, config backend.BuildConfig) (string
 		if imageID, err = squashBuild(build, b.imageComponent); err != nil {
 			return "", err
 		}
+	}
+
+	if options.Squash || useBuildKit {
```
Comment (tonistiigi) [inline api/server/backend/build/backend.go]: Why does this always runs with `useBuildKit`?

File: api/server/backend/build/backend.go [hunk 4]
```diff
@@ -72,8 +72,11 @@ func (b *Backend) Build(ctx context.Context, config backend.BuildConfig) (string
...
 		if config.ProgressWriter.AuxFormatter != nil {
-			if err = config.ProgressWriter.AuxFormatter.Emit(types.BuildResult{ID: imageID}); err != nil {
+			if err = config.ProgressWriter.AuxFormatter.Emit("", types.BuildResult{ID: imageID}); err != nil {
```
Comment (tonistiigi) [inline api/server/backend/build/backend.go]: Why no ID in here, and in `emitImageID`

Comment (tiborvass) [reply api/server/backend/build/backend.go]: @tonistiigi because that was the previous behavior: anything that used Emit, did not set the ID field.

**Comment by tonistiigi** [comment]
LGTM

File: builder/builder-next/builder.go [hunk 5]
```diff
@@ -258,25 +259,9 @@ func (b *Builder) Build(ctx context.Context, opt backend.BuildConfig) (*builder.
...
 			}
-
-			auxJSONBytes, err := json.Marshal(dt)
-			if err != nil {
+			if err := aux.Emit("moby.buildkit.trace", dt); err != nil {
 				return err
 			}
-			auxJSON := new(json.RawMessage)
-			*auxJSON = auxJSONBytes
-			msgJSON, err := json.Marshal(&jsonmessage.JSONMessage{ID: "moby.buildkit.trace", Aux: auxJSON})
```
Comment (thaJeztah) [inline builder/builder-next/builder.go]: We'll have to document the new format in the swagger.yaml before BuildKit leaves experimental

File: pkg/streamformatter/streamformatter.go [hunk 6]
```diff
@@ -139,14 +139,14 @@ type AuxFormatter struct {
...
-func (sf *AuxFormatter) Emit(aux interface{}) error {
+func (sf *AuxFormatter) Emit(id string, aux interface{}) error {
 	auxJSONBytes, err := json.Marshal(aux)
 	if err != nil {
 		return err
 	}
 	auxJSON := new(json.RawMessage)
 	*auxJSON = auxJSONBytes
-	msgJSON, err := json.Marshal(&jsonmessage.JSONMessage{Aux: auxJSON})
+	msgJSON, err := json.Marshal(&jsonmessage.JSONMessage{ID: id, Aux: auxJSON})
```
Comment (thaJeztah) [inline pkg/streamformatter/streamformatter.go]: Slightly wondering id `ID` as name may give the expectation it holds something globally unique, but don't have a different suggestion at this point ("type"? "class"? - not really better)

File: pkg/streamformatter/streamformatter_test.go [hunk 7]
```diff
@@ -106,7 +106,7 @@ func TestAuxFormatterEmit(t *testing.T) {
 	sampleAux := &struct {
 		Data string
 	}{"Additional data"}
-	err := aux.Emit(sampleAux)
+	err := aux.Emit("", sampleAux)
```
Comment (thaJeztah) [inline pkg/streamformatter/streamformatter_test.go]: This is a bit awkward (ID being optional)

**Review by thaJeztah — APPROVED** [review]
LGTM

left some notes for later consideration (it's still experimental, so there's likely other improvements we can make along the way)

File: pkg/streamformatter/streamformatter_test.go (hunk 7 above)
Comment (tiborvass) [reply pkg/streamformatter/streamformatter_test.go]: ID was added, most of the time ID is not set (for regular stream output). The whole streamformatter thing is awkward.

File: pkg/streamformatter/streamformatter.go (hunk 6 above)
Comment (tiborvass) [reply pkg/streamformatter/streamformatter.go]: It's been `ID` forever I don't think we'll change it now.

**Comment by tonistiigi** [comment]
@thaJeztah The ID of jsonmessage in the stream is already used to signify the grouping of multiple records, for example for updates that correspond to the same blob. The ID is optional as well atm. with only certain records that can be handled in a special way having that value.

**Comment by thaJeztah** [comment]
Yeah @tiborvass just pointed me to that; looks like a legacy we'll have to carry 😅

**Pull request merged**